flepimop/gempyor_pkg/src/gempyor/shared_cli.py

```diff
@@ -0,0 +1,288 @@
+
+    Returns:
+        The original function with an updated docstring.
+
+    Notes:
+        Adapted from https://stackoverflow.com/a/78533451.
+
+    Examples:
+        >>> import click
+        >>> opt = click.Option("-n", "--number", type=int, help="Your favorite number.")
+        >>> @click_helpstring(opt)
+        ... def test_cli(num: int) -> None:
+        ...     print(f"Your favorite number is {num}!")
+        ...
+        >>> print(test_cli.__doc__)
+
+
+        Command Line Interface arguments:
+
+        n: Int
+    """
+    if not isinstance(params, list):
+        params = [params]
+
+    def decorator(func):
+        # Generate the additional docstring with args from the specified functions
+        additional_doc = "\n\tCommand Line Interface arguments:\n"
+        for param in params:
+            paraminfo = param.to_info_dict()
+            additional_doc += (
+                f"\n\t\t{paraminfo['name']}: {paraminfo['type']['param_type']}\n"
+            )
+
+        if func.__doc__ is None:
+            func.__doc__ = ""
+        func.__doc__ += additional_doc
+
+        return func
+
+    return decorator
+
+
+# TODO: create a custom command decorator cls ala: https://click.palletsprojects.com/en/8.1.x/advanced/#command-aliases
+# to also apply the `@click_helpstring` decorator to the command. Possibly to also default the params argument, assuming
+# enough commands have consistent option set?
+
+mock_context = click.Context(
+    click.Command("mock"),
+    info_name="Mock context for non-click use of parse_config_files",
+)
+
+
+@click_helpstring([config_files_argument] + list(config_file_options.values()))
+def parse_config_files(
+    cfg: confuse.Configuration = config, ctx: click.Context = mock_context, **kwargs
+) -> confuse.Configuration:
+    """
+    Parse configuration file(s) and override with command line arguments
+
+    Args:
+        cfg: the configuration object to update; defaults to global `.utils.config`
+        ctx: the click context (used for type casting); defaults to a mock context for non-click use; pass actual context if using in a click command
+        **kwargs: see auto generated CLI items below. Unmatched keys will be ignored + a warning will be issued
+
+    Returns: None; side effect: updates the `cfg` argument
+    """
+
+    def _parse_option(param: click.Parameter, value: Any) -> Any:
+        """internal parser to autobox values"""
+        if (param.multiple or param.nargs == -1) and not isinstance(
+            value, (list, tuple)
+        ):
+            value = [value]
+        return param.type_cast_value(ctx, value)
+
+    parsed_args = {config_files_argument.name}.union(
+        {option.name for option in config_file_options.values()}
+    )
+
+    # warn re unrecognized arguments
+    if unknownargs := [
+        k for k in parsed_args.difference(kwargs.keys()) if kwargs.get(k) is not None
+    ]:
+        warnings.warn(f"Unused arguments: {unknownargs}")
```

Comment:
i think there's going to be a bit of iteration on this part of it - this is really just here for direct use rather than via the CLI, which probably wants some other feature iteration (e.g. how to handle defaults for click parameters for arguments not manually supplied).

i'm fine with as-is or minor tweaks, but i think this needs some user time to really figure out what's the right approach.